### DISCUSSION TITLE
Fragment component

### DISCUSSION THREAD

**Discussion by CyberAP** [vuejs/rfcs#448] [RFC Discussions]
This is a discussion for [Fragment component proposal](https://github.com/vuejs/rfcs/pull/449).

**Comment by jods4** [comment]
How is `h('fragment')` implemented? Would it be much more efficient than a userland functional component that performs this?
```js
// Code slightly simplified for clarity
function wrapper({ shouldWrap }, { slots }) {
  return shouldWrap 
    ? h("Wrapper", slots.default())
    : slots.default();
}
```

**Reply by CyberAP** [reply]
Good question!

How the code would look like if we'd wanted to pass a dynamic component with props there?

I assume if we update the implementation it would look like this:

```html
<wrapper :should-wrap="shouldWrap" :component="wrapperComponent" :attributes="attributes">
  Some contents...
</wrapper>
```

Which solves basic cases, but we're in fact re-implementing Vue's functionality here. If we'd like to add directives there it'll get more complicated.

**Reply by jods4** [reply]
For dynamic componenet + props I'd rather have gone something like below.
It should work for props at least, I am not sure what the behavior for directives is gonna be here.
```js
function wrapper({ shouldWrap, component }, { slots, attrs }) {
  return shouldWrap 
    ? h(component, attrs, slots.default())
    : slots.default();
}
```

How these kind of things work needs to be clearly spelled out in the RFC, as I assume you don't want wrapper properties (or directives?) to fallback inside the `h('fragment')` but that'd be the default behaviour unless specified otherwise?

The code above is a pretty simple functional component, at this stage I wouldn't say it's "re-implementing Vue" and should work except maybe with directives.

I haven't thought about directives deeply and they might justify something in Vue core. 
Otherwise I think this could easily be a npm package.

**Reply by posva** [reply]
As @jods4 pointed out [this can be easily implemented](https://sfc.vuejs.org/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcbiAgaW1wb3J0IEZyYWdtZW50IGZyb20gJy4vRnJhZ21lbnQudnVlJ1xuXG5jb25zdCBtc2cgPSByZWYoJ2hlbGxvJylcbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG5cdDxwPlxuICA8L3A+XG4gIDxpbnB1dCB2LW1vZGVsPVwibXNnXCI+XG4gIDxGcmFnbWVudCA6aXM9XCJtc2cgPT09ICdoZWxsbycgPyAnaDInIDogbnVsbFwiIHN0eWxlPVwiY29sb3I6IHJlZFwiPlxuICBcdHt7IG1zZyB9fVxuICA8L0ZyYWdtZW50PlxuPC90ZW1wbGF0ZT4iLCJpbXBvcnQtbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJ2dWVcIjogXCJodHRwczovL3NmYy52dWVqcy5vcmcvdnVlLnJ1bnRpbWUuZXNtLWJyb3dzZXIuanNcIlxuICB9XG59IiwiRnJhZ21lbnQudnVlIjoiPHNjcmlwdCBsYW5nPVwidHNcIj5cbmV4cG9ydCBkZWZhdWx0IHsgaW5oZXJpdEF0dHJzOiBmYWxzZSB9XG48L3NjcmlwdD5cblxuPHNjcmlwdCBzZXR1cCBsYW5nPVwidHNcIj5cbiAgZGVmaW5lUHJvcHM8eyBpczogYW55IH0+KClcbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIDxjb21wb25lbnQgdi1pZj1cImlzXCIgOmlzPVwiaXNcIiB2LWJpbmQ9XCIkYXR0cnNcIj5cbiAgICA8c2xvdCAvPlxuICA8L2NvbXBvbmVudD5cbiAgPHNsb3Qgdi1lbHNlIC8+XG48L3RlbXBsYXRlPiJ9) in userland. The RFC should probably go deeper and outline benefits or existing limitations to justify adding something new.

**Reply by CyberAP** [reply]
An additional custom directives example has been added to illustrate the problems that can not be solved with a user-land `fragment` component.

**Reply by posva** [reply]
directives work too

**Comment by unrevised6419** [comment]
Strange why `import { Fragment } from 'vue'` does not work by default 🤔 

<img width="528" alt="image" src="https://user-images.githubusercontent.com/1881266/168831656-b06b2a92-d6cd-4464-9132-6d5c87fd2f17.png">

**Reply by edison1105** [reply]
a fragment can only have array children since they are either generated by the compiler, or implicitly created from arrays. but slot is an object, so its not work.

**Comment by unrevised6419** [comment]
**Fragment.vue** 🤣
```vue
<template>
  <slot />
</template>
```

**Reply by cepm-nate** [reply]
Trying to use `import { Fragment } from 'vue';` gives me "Unknown custom element: <Fragment>...", even when within the 'components' array. But when I try to create my own "fragment.vue" like you show above (so simple!), Vue's ESLINT whines about "Cannot use <slot> as a component root element because it may contain multiple nodes".

I can disable the whining, but would rather follow the rules for future issues.

Does anyone have a working "implementation in userland" of this?

**Reply by unrevised6419** [reply]
What version of Vue are you using?

**Reply by cepm-nate** [reply]
Vue 2.6.14. Thankfully I can use some CSS selectors and abolish the styling of the extra wrapper element when it's not needed. Not a solution for everyone, but workable for me now. I still vote for this feature to be considered in the future. :-)

**Reply by unrevised6419** [reply]
Vue 2 does not support multiple root elements. I think you can try this. 🤔 
`display: contents;` will make in a way the element "virtual"

```vue
<template>
  <div style="display: contents;">
    <slot />
  </div>
</template>
```

https://developer.mozilla.org/en-US/docs/Web/CSS/display

**Reply by ferferga** [reply]
The problem with this approach is that attributes will not be passed through. Either the Slot API is redesigned as well to allow this, or a Fragment component, as proposed here, it's done.

The reason why it's not supported is due to [scoped slots](https://vuejs.org/guide/components/slots.html#scoped-slots)

**Comment by andrewsm80** [comment]
If something like this does eventually get implemented, I think it would make a lot more sense to use a directive, analogous to the `v-if` directive:

```
  <div v-wrap-if="someCondition">
    <span>
      This is a span that is sometimes wrapped up in a div.
    </span>
  </div>

  <WrapperComponent v-wrap-if="someCondition" :foo="getFoo()" :bar="getBar()">
    <span>
      This is a span that is sometimes wrapped up in a WrapperComponent.
      The getFoo() and getBar() expressions should only be evaluated if the wrapper component is rendered.
    </span>
  </div>

```

Conditionally rendering wrapper elements/components is such a basic operation, the syntax should match that in terms of simplicity.

**Comment by distor-sil3nt** [comment]
Is there any news on this? The conditional wrapper feature would be really useful and I'm wondering whether something was already implemented in this regard? Either as a custom directive or via the `<compontent>` tag.

**Comment by katerlouis** [comment]
Any new status update? :)

**Comment by distor-sil3nt** [comment]
What I did as a patch for this missing feature was to create a helper component / module in my utils directory (under `src/utils/helpers`) called `useFragment.ts` and basically just return the slots or default slot for this component:

`useFragment.ts`
```typescript
import { defineComponent } from 'vue'

export const useFragment = defineComponent({
  inheritAttrs: false,
  render() {
    return this.$slots && (typeof this.$slots.default === 'function' ? this.$slots.default() : '')
  },
})
```

In the template I used it like this:

```vue
<template>
  <Component :is="!isDocument ? 'a' : useFragment" :href="item.url" target="_blank">
    {{ item.title }}
  </Component>
</template>

<script setup lang="ts">
   import { useFragment } from '@/utils'
   
   [...]
</script>
```

I hope this helps someone looking for a solution to this. It would still be better if it was integrated as a default functionality in Vue. I would highly appreciate it if someone from the core team could attend to this matter.

**Reply by katerlouis** [reply]
sleek workaround; wondering if Vue will come up with a native solution

**Comment by katerlouis** [comment]
Inspired by @distor-sil3nt, I made a Nuxt 2 (Vue 2.7) compatible component for optional wrapping:

```vue
<script>
export default {
  name: 'Fragment',

  render(h) {
    if (this.$slots && this.$slots.default) {
      return this.$slots.default;
    }

    return null;
  },
};
</script>
```

usage:
```html
<component :is="isLink ? 'a' : 'Fragment' :href="url">Click me</component>
```

**Comment by slavco86** [comment]
Looks like a neat solution in Vue 3
|https://dev.to/alexander-nenashev/conditional-wrap-component-in-vue-3-pt3-8dp
